tests/components/reolink/test_init.py

```diff
@@ -0,0 +1,114 @@
+"""Test the Reolink init."""
+from typing import Any
+from unittest.mock import AsyncMock, MagicMock, Mock
+
+import pytest
+from reolink_aio.exceptions import ReolinkError
+
+from homeassistant.components.reolink import const
+from homeassistant.config import async_process_ha_core_config
+from homeassistant.config_entries import ConfigEntryState
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers import issue_registry as ir
+
+from tests.common import MockConfigEntry
+
+pytestmark = pytest.mark.usefixtures("reolink_connect", "reolink_platforms")
+
+
+@pytest.mark.parametrize(
+    ("attr", "value", "expected"),
+    [
+        (
+            "is_admin",
+            False,
+            ConfigEntryState.SETUP_ERROR,
+        ),
+        (
+            "get_host_data",
+            AsyncMock(side_effect=ReolinkError("Test error")),
+            ConfigEntryState.SETUP_RETRY,
+        ),
+        (
+            "get_host_data",
+            AsyncMock(side_effect=ValueError("Test error")),
+            ConfigEntryState.SETUP_ERROR,
+        ),
+        (
+            "get_states",
+            AsyncMock(side_effect=ReolinkError("Test error")),
+            ConfigEntryState.SETUP_RETRY,
+        ),
+        (
+            "supported",
+            Mock(return_value=False),
+            ConfigEntryState.LOADED,
+        ),
+        (
+            "check_new_firmware",
+            AsyncMock(side_effect=ReolinkError("Test error")),
+            ConfigEntryState.LOADED,
+        ),
+    ],
+)
+async def test_failures_parametrized(
+    hass: HomeAssistant,
+    reolink_connect: MagicMock,
+    config_entry: MockConfigEntry,
+    attr: str,
+    value: Any,
+    expected: ConfigEntryState,
+) -> None:
+    """Test outcomes when changing errors."""
+    setattr(reolink_connect, attr, value)
+    assert await hass.config_entries.async_setup(config_entry.entry_id) is (
+        expected == ConfigEntryState.LOADED
+    )
+    await hass.async_block_till_done()
+
+    assert config_entry.state == expected
+
+
+async def test_update_listener(
+    hass: HomeAssistant, config_entry: MockConfigEntry
+) -> None:
+    """Test the update listener."""
+    assert await hass.config_entries.async_setup(config_entry.entry_id)
+    await hass.async_block_till_done()
+
+    assert config_entry.title == "test_reolink_name"
+
+    hass.config_entries.async_update_entry(config_entry, title="New Name")
+    await hass.async_block_till_done()
+
+    assert config_entry.title == "New Name"
```

Comment:
This is not testing the listeners of the integration, this is just testing `async_update_entry`

Reply:
I have changed the name of the function to `test_async_update_entry` and updated the docstring

Reply:
That doesn't change anything. The comment is about this is not testing integration details, but Home Assistant itself.

Unresolving the conversation.

Reply:
So how do I get coverage of the `entry_update_listener` and `async_unload_entry` functions inside __init__.py?
https://github.com/home-assistant/core/blob/85bcf11aeb8e7af8045ca2081411b548c8e8a7a1/homeassistant/components/reolink/__init__.py#L144-L160

Could you point to an example of another component that covers those with a test?

Reply:
So, with the current test, you can remove that function completely, and the test will still pass.

I think you should test if `stop()` has been called. That would put an assertion in place that ensures this code runs.

Reply:
I get it, thank you for your explantion, will work on it now!

Reply:
I am now checking `assert reolink_connect.logout.call_count == 0` before reloading and then a`ssert reolink_connect.logout.call_count == 1` after reloading.

host.stop will call host.disconnect that will call host._api.logout which I am checking.

Reply:
Hmm interesting, that means `host` isn't mocked? Should it be mocked?

Reply:
`host` lives within HomeAssistant in the `host.py` file, I don't think it schould be mocked, only the underlying functions of host from the client library and webhook component schould be mocked in my view.

At some point I schould remove `host.py` from .coveragerc since a large part is already covered by the tests of the config_flow and init files.
But then I schould add some additional tests to bring the coverage to 100% of the `host.py` file.

I propose we do that in a seprate PR.

Reply:
Current coverage of host.py is 62%

Reply:
I see now, ok alright, looks good to me 👍